Add Config Flow to Obihai

ejpenney opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
N/A

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Adds config_flow to the Obihai integration, deprecates YAML configuration.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [x] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes: N/A
- This PR is related to issue: N/A
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/26354

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [[www.home-assistant.io][docs-repository](http://www.home-assistant.io][docs-repository/)]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

dshokouhi commented:
This great! Please also consider adding yourself to the codeowners for this integration as I have not done any updates to this in a very long time 😬

ejpenney commented:
I don't know what the heck is going on.  These tests are passing for me offline.

ejpenney commented:
@dshokouhi, looks like I got the pre-commit working. Added myself as code owner. Let me know if you think anything else needs done!

epenet reviewed (changes requested):
Please split into smaller pr
- one for code owners 
- one for reboot service 
- one for ConfigFlow (without dhcp)
- one follow up for dhcp

ejpenney committed:
Obihai: Config Flow Only

ejpenney committed:
Remove reboot service

epenet commented on .coveragerc:
```diff
@@ -803,7 +803,9 @@ omit =
     homeassistant/components/nx584/alarm_control_panel.py
     homeassistant/components/oasa_telematics/sensor.py
-    homeassistant/components/obihai/*
+    homeassistant/components/obihai/const.py
```
```suggestion
```

epenet commented on homeassistant/components/obihai/__init__.py:
```diff
@@ -1 +1,36 @@
+    return await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
+
+
+async def async_reload_entry(hass: HomeAssistant, entry: ConfigEntry) -> None:
```
Can you keep options flow for a follow-up PR?

epenet commented on homeassistant/components/obihai/__init__.py:
```diff
@@ -1 +1,36 @@
+from .const import CONF_OBIHAI_HOST, PLATFORMS
+
+__all__ = [
+    "CONF_OBIHAI_HOST",
```
I don't think this is needed

epenet reviewed (changes requested):
Also please ensure that you have 100% test coverage on config flow.

ejpenney committed:
Update .coveragerc

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

ejpenney committed:
PR Feedback

epenet commented on homeassistant/components/obihai/sensor.py:
```diff
@@ -35,46 +31,47 @@
     discovery_info: DiscoveryInfoType | None = None,
 ) -> None:
     """Set up the Obihai sensor platform."""
+    LOGGER.warning(
```
Can you use the issue registry instead?

epenet commented on homeassistant/components/obihai/config_flow.py:
```diff
@@ -0,0 +1,118 @@
+            return self.async_create_entry(
+                title=user_input.get(CONF_NAME, user_input[CONF_OBIHAI_HOST]),
+                data={},
+                options={
```
Why use options and not data?
I don't think it should be possible to change yhe host; and username/password should be handled via reauth.

epenet commented on homeassistant/components/obihai/strings.json:
```diff
@@ -0,0 +1,36 @@
+      "already_configured": "[%key:common::config_flow::abort::already_configured_device%]"
+    }
+  },
+  "options": {
```
Not required

epenet replied:
`options` still needs to be removed from `strings.json`

ejpenney committed:
Use Issue Registry

ejpenney committed:
Add config_flow tests

ejpenney committed:
Another pass with pre-commit

ejpenney commented:
@epenet, I believe I have addressed all your concerns.  When this gets submitted I'll submit a few other PRs.  The other config_flow changes might take a bit since they don't have tests...

ejpenney commented on homeassistant/components/obihai/obihai_api.py:
```diff
@@ -0,0 +1,166 @@
+        return True
+
+
+class ObihaiServiceSensors(SensorEntity):
```
Moving this class to Sensor creates cyclical imports that fail the test_config_flow test set.

ejpenney replied:
ok, I think I got it.  Just moved both the Sensor class and the Connection class (really just a collection of sensors anyways) to sensor.py and the import error went away.

ejpenney committed:
Resolve cyclical import and move sensorClasses to sensor

epenet commented on homeassistant/components/obihai/config_flow.py:
```diff
@@ -0,0 +1,116 @@
+
+    VERSION = 1
+
+    def __init__(self) -> None:
+        """Initialize."""
+        self._host: str | None = None
+
```
`self._host` is never set
```suggestion
```

epenet commented on homeassistant/components/obihai/config_flow.py:
```diff
@@ -0,0 +1,116 @@
+from .obihai_api import validate_auth
+
+
+@callback
+def async_get_schema(
+    defaults: dict[str, Any] | MappingProxyType[str, Any]
+) -> vol.Schema:
+    """Return Obihai schema."""
+    schema = {
+        vol.Required(CONF_HOST, default=defaults.get(CONF_OBIHAI_HOST, "")): str,
+        vol.Optional(
+            CONF_USERNAME,
+            description={
+                "suggested_value": defaults.get(CONF_USERNAME, DEFAULT_USERNAME)
+            },
+        ): str,
+        vol.Optional(
+            CONF_PASSWORD,
+            default=defaults.get(CONF_PASSWORD, DEFAULT_PASSWORD),
+        ): str,
+    }
+
+    return vol.Schema(schema)
```
It is easier to make the schema a constant (with system defaults), and then use `self.add_suggested_values_to_schema`

See
https://github.com/home-assistant/core/blob/cab8a59be48a027e2da9ed96aa782f7018d91b59/homeassistant/components/sfr_box/config_flow.py#L61

epenet commented on homeassistant/components/obihai/config_flow.py:
```diff
@@ -0,0 +1,116 @@
+            data_schema=async_get_schema(user_input),
+        )
+
+    async def async_validate_input(self, user_input: dict[str, Any]) -> FlowResult:
```
Maybe it's just me, but I think it is not useful to extract this to a separate method.
You can just use:
```python
    errors: dict[str, str] = {}
    if user_input is not None and not (errors := await async_validate_credentials(self.hass, user_input)):
        self._async_abort_entries_match(...)
        return self.async_create_entry(...)

    data_schema = self.add_suggested_values_to_schema(DATA_SCHEMA, user_input)
    return self.async_show_form(...)
```

ejpenney replied:
It will be when we add the OptionsFlow back in, as that structure will need to validate input as well.  Implementing it this way will result in duplicate code or unnecessary increase to churn on this file.

ejpenney replied:
Alright, I'll just do it and deal with the fallout in the next PR.  Would really like to get this merged, seeing as it's not really the feature I wanted to add, but those two items are dependent on this.

epenet commented on homeassistant/components/obihai/const.py:
```diff
@@ -0,0 +1,16 @@
+
+LOGGER = logging.getLogger(__package__)
+
+CONF_OBIHAI_HOST: Final = "host"
```
Why not use the global `CONF_HOST` constant?

epenet commented on homeassistant/components/obihai/obihai_api.py:
```diff
@@ -0,0 +1,31 @@
+"""Support for Obihai Sensors."""
```
I think you should rename this file - we know it's `obihai` already from the domain.
```suggestion
"""Support for Obihai connectivity."""
```

epenet commented on homeassistant/components/obihai/sensor.py:
```diff
@@ -146,19 +187,25 @@ def icon(self):
         return "mdi:phone"
 
-    def update(self) -> None:
+    def update(self) -> bool:
```
Why return a bool?
```suggestion
    def update(self) -> None:
```

epenet commented on homeassistant/components/obihai/sensor.py:
```diff
@@ -146,19 +187,25 @@ def icon(self):
+    def update(self) -> bool:
         """Update the sensor."""
-        services = self._pyobihai.get_state()
+        if self._pyobihai.check_account():
```
You can return early here instead, and un-indent:
```suggestion
        if not self._pyobihai.check_account():
            self._state = None
            return
```

epenet commented on homeassistant/components/obihai/sensor.py:
```diff
@@ -35,46 +32,90 @@
+        for key in services:
+            self.sensors.append(ObihaiServiceSensors(pyobihai, serial, key))
 
-    for key in call_direction:
-        sensors.append(ObihaiServiceSensors(pyobihai, serial, key))
+        if line_services is not None:
+            for key in line_services:
+                self.sensors.append(ObihaiServiceSensors(pyobihai, serial, key))
 
-    add_entities(sensors)
+        for key in call_direction:
+            self.sensors.append(ObihaiServiceSensors(pyobihai, serial, key))
```
I think these 7 lines should be moved to `async_setup_entry`, replacing `sensors = requester.sensors`
Then you can move `ObihaiConnection` to the api module, and expose `serial`/`services`/`line_services`/`call_direction` as properties

epenet commented on tests/components/obihai/test_config_flow.py:
```diff
@@ -0,0 +1,68 @@
+from . import USER_INPUT
+
+
+async def test_user_form(hass: HomeAssistant) -> None:
```
```suggestion
async def test_user_form(hass: HomeAssistant, mock_setup_entry: AsyncMock) -> None:
```

epenet commented on tests/components/obihai/test_config_flow.py:
```diff
@@ -0,0 +1,68 @@
+    assert result["type"] == FlowResultType.FORM
+    assert result["errors"] is None
+
+    with patch("pyobihai.PyObihai.check_account"), patch(
+        "homeassistant.components.obihai.async_setup_entry"
+    ) as mock_setup_entry:
```
Please take a look at https://github.com/home-assistant/core/pull/88661
I think it's better to ensure `async_setup_entry` is always mocked in config flow tests (with `pytestmark = pytest.mark.usefixtures`), and then inject the fixture in the test definition.

ejpenney committed:
Update homeassistant/components/obihai/config_flow.py

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

ejpenney committed:
Update homeassistant/components/obihai/sensor.py

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

ejpenney committed:
Update homeassistant/components/obihai/sensor.py

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

ejpenney committed:
Update tests/components/obihai/test_config_flow.py

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

ejpenney committed:
Another round of feedback

ejpenney committed:
More PR feedback

ejpenney commented:
@epenet, here for you again.

ejpenney commented on homeassistant/components/obihai/config_flow.py:
```diff
@@ -0,0 +1,90 @@
+    # DEPRECATED
+    async def async_step_import(self, config: dict[str, Any]) -> FlowResult:
+        """Handle a flow initialized by importing a config."""
+        self._async_abort_entries_match({CONF_HOST: config[CONF_HOST]})
```
@epenet _async_abort_entries states that I must specify already_configured in my string.json.  And indeed, my offline testing indicates this particular error message fails to display properly if it's missing from strings.json.  I suspect this should be added back for this PR.

![image](https://user-images.githubusercontent.com/21226768/221039314-00e1a7d3-5c1d-430f-a31c-5ef971b9f511.png)

epenet replied:
Yes you can add it. My main concern was to remove `options` section from the json file

ejpenney committed:
Offline testing, already_configured is required

epenet commented on homeassistant/components/obihai/config_flow.py:
```diff
@@ -0,0 +1,90 @@
+
+DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_HOST, default=""): str,
```
You don't need empty default 
```suggestion
        vol.Required(CONF_HOST): str,
```

epenet commented on homeassistant/components/obihai/config_flow.py:
```diff
@@ -0,0 +1,90 @@
+            self._async_abort_entries_match({CONF_HOST: user_input[CONF_HOST]})
+            return self.async_create_entry(
+                title=user_input.get(CONF_NAME, user_input[CONF_HOST]),
+                data={
```
Could this not be simplified to just pass the input?
```suggestion
                data=user_input
```

epenet commented on homeassistant/components/obihai/config_flow.py:
```diff
@@ -0,0 +1,90 @@
+        ):
+            self._async_abort_entries_match({CONF_HOST: user_input[CONF_HOST]})
+            return self.async_create_entry(
+                title=user_input.get(CONF_NAME, user_input[CONF_HOST]),
```
NAME is not part of the schema 
```suggestion
                title=user_input[CONF_HOST],
```

epenet commented on homeassistant/components/obihai/sensor.py:
```diff
@@ -35,46 +32,61 @@
 
-def setup_platform(
+# DEPRECATED
+async def async_setup_platform(
```
I don't think you need to make this async anymore

ejpenney replied:
Does it hurt to have it run async?  Just makes sure the Obihai setup isn't blocking, and it goes away after the deprecation sunset...

epenet commented on homeassistant/components/obihai/sensor.py:
```diff
@@ -162,3 +178,5 @@ def update(self) -> None:
         if self._service_name in call_direction:
             self._state = call_direction.get(self._service_name)
+
+        self._state = None
```
I think this line shouldn't be here. Leftover?

ejpenney replied:
Thanks, I think you're right.

ejpenney committed:
Update homeassistant/components/obihai/config_flow.py

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

ejpenney committed:
Update homeassistant/components/obihai/config_flow.py

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

ejpenney committed:
Update homeassistant/components/obihai/config_flow.py

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

ejpenney committed:
Cleanup

ejpenney commented:
@epenet, are we good?

epenet commented on homeassistant/components/obihai/config_flow.py:
```diff
@@ -0,0 +1,86 @@
+    ) -> FlowResult:
+        """Handle a flow initialized by the user."""
+        errors: dict[str, str] = {}
+        if user_input is not None and not (
+            errors := await async_validate_credentials(self.hass, user_input)
+        ):
+            self._async_abort_entries_match({CONF_HOST: user_input[CONF_HOST]})
```
Maybe abort should be done before validate_credentials.
There is no point validating the credentials if it already exists...

epenet commented on homeassistant/components/obihai/sensor.py:
```diff
@@ -35,46 +32,61 @@
+        translation_key="manual_migration",
+    )
+
+    if discovery_info:
+        config = PLATFORM_SCHEMA(discovery_info)
```
I am a little confused by this line.
When does it get used?

epenet commented on homeassistant/components/obihai/config_flow.py:
```diff
@@ -0,0 +1,86 @@
+)
+
+
+async def async_validate_credentials(
+    hass: HomeAssistant, user_input: dict[str, Any]
+) -> dict[str, str]:
+    """Manage Obihai options."""
+    errors = {}
+    result = await hass.async_add_executor_job(
+        validate_auth,
+        user_input[CONF_HOST],
+        user_input[CONF_USERNAME],
+        user_input[CONF_PASSWORD],
+    )
+
+    if not result:
+        errors["base"] = "cannot_connect"
+
+    return errors
```
I still think this whole method should be simplified to only return a bool, and the errors should be set within the flow.

```suggestion
async def async_validate_credentials(
    hass: HomeAssistant, user_input: dict[str, Any]
) -> bool:
    """Validate Obihai credentials."""
    return await hass.async_add_executor_job(
        validate_auth,
        user_input[CONF_HOST],
        user_input[CONF_USERNAME],
        user_input[CONF_PASSWORD],
    )
```

epenet replied:
At this stage you have not yet convinced me.
You can always re-adjsut later when the extra entry-points are added and you have a real use case.

ejpenney replied:
Guess I don't have much of a choice.  In your court again.

ejpenney committed:
Update homeassistant/components/obihai/__init__.py

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

ejpenney committed:
PR feedback

ejpenney commented on homeassistant/components/obihai/__init__.py:
```diff
@@ -1 +1,18 @@
+    """Set up from a config entry."""
+    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
+
+    return True
```
I'm backing out this change, async_forward_entry_setups doesn't return a value, and async_setup_entry must return bool (according to mypy).  So this change does not work.

ejpenney committed:
Backout today's changes: Fix mypy error

epenet commented on homeassistant/components/obihai/connectivity.py:
```diff
@@ -0,0 +1,66 @@
+
+    def update(self) -> bool:
+        """Validate connection and retrieve a list of sensors."""
+        self.pyobihai = get_pyobihai(self.host, self.username, self.password)
```
Should this not check if pyobihai already exists?
```suggestion
        if not self.pyobihai:
            self.pyobihai = get_pyobihai(self.host, self.username, self.password)
```

And maybe `check_account` can also be guarded?

epenet commented on tests/components/obihai/test_config_flow.py:
```diff
@@ -0,0 +1,70 @@
+        )
+        await hass.async_block_till_done()
+
+    assert result["errors"]["base"] == "cannot_connect"
```
```suggestion
    assert result["type"] == FlowResultType.FORM
    assert result["step_id"] == "user"
    assert result["errors"]["base"] == "cannot_connect"
```

epenet commented on tests/components/obihai/test_config_flow.py:
```diff
@@ -0,0 +1,70 @@
+
+import pytest
+
+from homeassistant import config_entries, data_entry_flow
```
```suggestion
from homeassistant import config_entries
```

epenet commented on tests/components/obihai/test_config_flow.py:
```diff
@@ -0,0 +1,70 @@
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+    assert result["type"] == FlowResultType.FORM
```
```suggestion
    assert result["type"] == FlowResultType.FORM
    assert result["step_id"] == "user"
```

epenet commented on tests/components/obihai/test_config_flow.py:
```diff
@@ -0,0 +1,70 @@
+        )
+        await hass.async_block_till_done()
+
+    assert result["type"] == data_entry_flow.FlowResultType.CREATE_ENTRY
```
```suggestion
    assert result["type"] == FlowResultType.CREATE_ENTRY
```

ejpenney committed:
Don't plan ahead

ejpenney committed:
Update tests/components/obihai/test_config_flow.py

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

ejpenney committed:
Update tests/components/obihai/test_config_flow.py

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

ejpenney committed:
Update tests/components/obihai/test_config_flow.py

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

ejpenney committed:
Update tests/components/obihai/test_config_flow.py

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

ejpenney committed:
PR feedback

epenet commented on homeassistant/components/obihai/strings.json:
```diff
@@ -0,0 +1,26 @@
+      }
+    },
+    "error": {
+      "cannot_connect": "[%key:common::config_flow::error::cannot_connect%]",
+      "invalid_auth": "[%key:common::config_flow::error::invalid_auth%]"
```
You are only using `cannot_connect` currently. Can you remove `invalid_auth` from strings.json or adjust the code accordingly?

epenet commented on homeassistant/components/obihai/config_flow.py:
```diff
@@ -0,0 +1,74 @@
+
+        if user_input is not None:
+            self._async_abort_entries_match({CONF_HOST: user_input[CONF_HOST]})
+            if not await self.hass.async_add_executor_job(
+                validate_auth,
+                user_input[CONF_HOST],
+                user_input[CONF_USERNAME],
+                user_input[CONF_PASSWORD],
+            ):
+                errors["base"] = "cannot_connect"
+            else:
+                return self.async_create_entry(
+                    title=user_input[CONF_HOST],
+                    data=user_input,
+                )
```
You can invert the logic here:
```suggestion
            if await self.hass.async_add_executor_job(
                validate_auth,
                user_input[CONF_HOST],
                user_input[CONF_USERNAME],
                user_input[CONF_PASSWORD],
            ):
                return self.async_create_entry(
                    title=user_input[CONF_HOST],
                    data=user_input,
                )
            errors["base"] = "cannot_connect"
```

ejpenney committed:
Update homeassistant/components/obihai/config_flow.py

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

ejpenney committed:
Cleanup strings

epenet reviewed (approved):
LGTM 👍

MartinHjelmare commented on homeassistant/components/obihai/config_flow.py:
```diff
@@ -0,0 +1,73 @@
+    async def async_step_import(self, config: dict[str, Any]) -> FlowResult:
+        """Handle a flow initialized by importing a config."""
+        self._async_abort_entries_match({CONF_HOST: config[CONF_HOST]})
+        return self.async_create_entry(
```
We should test the connection to the device with the config before creating the config entry, as we do in the user step. Abort if it doesn't work.

ejpenney replied:
Would you have objections to just using LOGGER to drop a warning in system log if YAML import failed?  I suppose we could create an "issue", but that seems overkill IMHO.

ejpenney replied:
Strike that, I'm just going to use async_abort and hand it cannot_connect.

MartinHjelmare commented on homeassistant/components/obihai/sensor.py:
```diff
@@ -35,46 +32,58 @@
+        "manual_migration",
+        breaks_in_ha_version="2023.6.0",
+        is_fixable=False,
+        severity=issue_registry.IssueSeverity.ERROR,
```
Please use warning severity.

MartinHjelmare commented on homeassistant/components/obihai/sensor.py:
```diff
@@ -148,6 +157,10 @@ def icon(self):
 
     def update(self) -> None:
         """Update the sensor."""
+        if not self._pyobihai.check_account():
```
What does `False` mean here? If we can't connect we should set the `available` property to `False` in the entity. If credentials are bad we should start a reauth flow.

https://developers.home-assistant.io/docs/config_entries_config_flow_handler#reauthentication

ejpenney replied:
The `available` property will return False if `self._state is None` which is what we've done here.  I'll take a look at the re-auth feedback, though I think that should be a PR in-and-of-itself as it would require new classes and tests.

MartinHjelmare replied:
Setting the native value to `None` means that the state is unknown. This is different than state unavailable. It would be better not to couple these things to avoid confusion when reading the code.

MartinHjelmare commented on homeassistant/components/obihai/strings.json:
```diff
@@ -0,0 +1,25 @@
+  },
+  "issues": {
+    "manual_migration": {
+      "title": "Manual migration required for Obihai",
```
What manual migration is required?

ejpenney replied:
Can you be more specific about what's confusing here?  This is C+P from other components that require migration from YAML.  This is just a title.  The description is... more descriptive of the process.  Would you like the title to describe how to migrate?

frenck replied:
The title is incorrect. There is no manual migration required (it has already migrated). It should state the YAML configuration is deprecated and should be removed.

MartinHjelmare reviewed (commented):
Please address the comments in a new PR. Thanks!
